### ISSUE TITLE
MSE documentation is weak

### ISSUE THREAD

**Issue by Atcold** [pytorch/pytorch#88327]
### 📚 The doc issue

Currently we have the MSE described as following:

> `CLASS` `torch.nn.MSELoss`(`size_average=None`, `reduce=None`, `reduction='mean'`)[[SOURCE]](https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html#MSELoss)
Creates a criterion that measures the mean squared error (squared L2 norm) between each element in the input $x$ and target $y$.
> 
> The unreduced (i.e. with `reduction` set to `'none'`) loss can be described as:
>
> $$
> \ell(x, y) = L = \lbrace l_1,\dots,l_N\rbrace^\top, \quad l_n = \left( x_n - y_n \right)^2,
> $$

Which is not the case.
Say `x` and `y` have `N = 3` elements, `L` does **not** contain the `N` squared L2 norms.

```python
>>> # Define cost
>>> C = nn.MSELoss(reduction='none')
>>> x = torch.randn(3, 5)
>>> y = torch.randn(3, 5)
>>> C(x, y).size()
torch.Size([3, 5])
```

We're assuming the first dimension is the batch size, yes?

**Comment by albanD** [comment]
> We're assuming the first dimension is the batch size, yes?

Yes we always do that unless explicitly specified.

This looks accurate to me though? `x_n` is a 1D vector of size 5 in your example?

**Comment by Atcold** [comment]
> This looks accurate to me though? `x_n` is a 1D vector of size 5 in your example?

Correct, and the squared L2 distance between `x_n` and `y_n` is the scalar `(x_n - y_n).pow(2).sum()`.

**Comment by albanD** [comment]
Ho right, I think that this is not what this is computing then. This is just the element-wise L2 distance.
We can definitely update the doc to reflect that.

**Comment by Atcold** [comment]
Unless the element-wise L2 distance is not the correct thing to compute here.
The MSE is the average (across the batch) squared error (or distance).
So, even when choosing `'mean'` we should be summing across the non-batch dimensions, right?

**Comment by lezcano** [comment]
We usually call that  `batchmean` in the loss functions that support that. See e.g. https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html. `mean` (and `sum` for the matter) always refers to reducing the whole tensor to one number, that is, treating the input tensors as if they were vectors.

**Comment by Atcold** [comment]
I can see how an MSE between images would favour a mean over the $(r, g, b)$ squared differences rather than a sum over components. Hmm…

*Context: I'm writing a book and I'm getting some misalignment between the maths and the implementation.*

Anyhow, the doc should explain what the function is doing. Right now it's incorrectly doing so.

**Comment by lezcano** [comment]
I agree, there should be a centralised place where the notation used to describe loss functions is explained.

**Comment by Temirkul** [comment]
2 years later and the documentation is still not updated. Can you please update the documentation? It is confusing in its current form.
